`az --version`: fail-fast on connectivity problems

tjprescott opened the pull request:
Fixes #8762.

Adds a utility method `check_connectivity` to speed up the process of timing out on connectivity problems. Previously we just sent the request and waited for it to fail, but this proved far too slow. This method is designed to fail fast by default (though a connectivity failure is still slower than a success).

---

This checklist is used to make sure that common guidelines for a pull request are followed.

- [N/A] The PR has modified HISTORY.rst describing any customer-facing, functional changes. Note that this does not include changes only to help content. (see [Modifying change log](https://github.com/Azure/azure-cli/tree/master/doc/authoring_command_modules#modify-change-log)).

- [X] I adhere to the [Command Guidelines](https://github.com/Azure/azure-cli/blob/dev/doc/command_guidelines.md).

limingu commented on src/azure-cli-core/azure/cli/core/util.py:
```diff
@@ -459,3 +460,22 @@ def find_child_collection(parent, *args, **kwargs):
     return collection
+
+
+def check_connectivity(url='https://example.org', max_retries=0, timeout=1):
```
Is https://example.org reliable? Should we use azure.microsoft.com or other more reliable url to test the connectivity?

marstr commented on src/azure-cli-core/azure/cli/core/util.py:
```diff
@@ -88,23 +88,24 @@ def _update_latest_from_pypi(versions):
-                versions[mod]['pypi'] = version
-            except KeyError:
-                pass
-    except CalledProcessError:
```
IMHO, I would keep this try/catch for pip failure even if we've added a fail-fast above. It'll keep the robust behavior before, with the improved fail-fast. I think the counter argument, which I can't really speak to, is that the Python idiom would be to have the exception propagate up instead of returning a bool to indicate that the check was unsuccessful.

marstr commented:
Just to capture this thought somewhere: This change makes us more sensitive to transient network problems than we previously were, with the benefit of not hanging during a version check. For just about any other function, we would want to continue being robust. So, this behavior shouldn't be copied by future command authors. I do think it is the right thing to do in this instance though.

tjprescott commented:
@marstr comments addressed. Also, I set the default on max_retries to something more generally useful and then simply override it where I use it to be 0. This partially addressed your comment on transient network issues, which I agree with.
